File: sdk/python/src/fyuz/distributor.py
"""The ``client.distributor`` namespace — on-chain leaderboard fee distribution.

Fyuz streams a 0.3% leaderboard fee into a Distributor contract. Every round it
pays 90% of the pot pro-rata by points and hands the remainder to a
Chainlink-VRF-picked winner. These endpoints expose that ledger.

Every ``*_wei`` value returned here is an exact decimal **string**. Do not parse
one into a ``float`` — the values exceed the range a float represents exactly.
:func:`fyuz.to_decimal` and :func:`fyuz.wei_to_bnb` convert losslessly.
"""

from __future__ import annotations

from typing import List, Optional

from ._convert import parse_array, parse_object
from ._http import Transport, encode_path_segment
from .models import (
    AddressPayouts,
    Claimable,
    PayoutStats,
    Pot,
    RoundDetail,
    RoundReceipt,
    Shares,
)

__all__ = ["DistributorAPI"]


class DistributorAPI:
    """Distributor endpoints, reached through :attr:`fyuz.FyuzClient.distributor`.

    Example::

        with FyuzClient() as fyuz:
            pot = fyuz.distributor.get_pot()
            if pot.pot_bnb is None:
                print("pot unknown right now")   # NOT the same as 0
            else:
                print(f"{pot.pot_bnb} BNB up for grabs")
    """

    def __init__(self, transport: Transport) -> None:
        """Bind the namespace to a client's transport. Constructed by the client."""
        self._transport = transport

    def get_stats(self) -> PayoutStats:
        """Lifetime payback totals.

        Returns:
            Totals distributed and round count. ``total_paid_wei`` and
            ``largest_payout_wei`` are exact wei decimal strings.
        """
        payload = self._transport.request("GET", "/distributor/stats")
        return parse_object(payload, PayoutStats.from_dict)

    def get_pot(self) -> Pot:
        """The live undistributed pot and the round parameters.

        Returns:
            Pot size and schedule. ``pot_bnb`` and ``total_points`` are ``None``
            when unknown (RPC unreachable or no Distributor configured) — hide
            the figure rather than showing ``0``.
        """
        payload = self._transport.request("GET", "/distributor/pot")
        return parse_object(payload, Pot.from_dict)

    def get_shares(
        self,
        *,
        from_ts: Optional[int] = None,
        to_ts: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> Shares:
        """The share allocation the round-runner will post on-chain.

        Args:
            from_ts: Window start, UNIX seconds (``from`` on the wire).
                Defaults to the end of the last paid round.
            to_ts: Window end, UNIX seconds (``to`` on the wire). Defaults to
                now. Must be after ``from_ts`` or the server answers 400.
            limit: Max holders, top-N by points. Clamped server-side to 1-100
                (the contract's ``MAX_HOLDERS``). Default 100.

        Returns:
            Per-address shares plus ``packed``, the exact calldata for
            ``Distributor.postShares``.

        Raises:
            FyuzAPIError: 400 when ``to_ts`` is not after ``from_ts``.
        """
        payload = self._transport.request(
            "GET",
            "/distributor/shares",
            params={"from": from_ts, "to": to_ts, "limit": limit},
        )
        return parse_object(payload, Shares.from_dict)

    def list_rounds(
        self, *, limit: Optional[int] = None, offset: Optional[int] = None
    ) -> List[RoundReceipt]:
        """The public payout ledger, newest round first.

        Args:
            limit: Rows to return, clamped server-side to 1-100. Default 25.
            offset: Rows to skip. Default 0.

        Returns:
            Round receipts. Null wei fields mean the settlement has not been
            indexed yet, not that zero was paid.
        """
        payload = self._transport.request(
            "GET", "/distributor/rounds", params={"limit": limit, "offset": offset}
        )
        return parse_array(payload, RoundReceipt.from_dict)

    def get_round(self, round_id: int) -> RoundDetail:
        """One round's full receipt, including every recipient line.

        Args:
            round_id: On-chain round id.

        Returns:
            The receipt fields (flattened at the top level of the JSON) plus
            ``payouts``, richest first.

        Raises:
            FyuzNotFoundError: no such round.
        """
        path = f"/distributor/rounds/{encode_path_segment(round_id)}"
        return parse_object(self._transport.request("GET", path), RoundDetail.from_dict)

    def get_payouts(self, address: str) -> AddressPayouts:
        """Every payout a single wallet has received.

        Args:
            address: Wallet address.

        Returns:
            Lifetime total (exact wei string) and the individual lines, newest
            round first.
        """
        path = f"/distributor/payouts/{encode_path_segment(address)}"
        return parse_object(self._transport.request("GET", path), AddressPayouts.from_dict)

    def get_claimable(self, address: str) -> Claimable:
        """Wei owed to an address from payout pushes the contract could not complete.

        Withdrawn by calling ``claim()`` on the Distributor — the wallet signs,
        this API only reads.

        Args:
            address: Wallet address.

        Returns:
            The pending amount. ``claimable_wei`` is ``None`` when unknown —
            hide the banner rather than saying nothing is owed.
        """
        path = f"/distributor/claimable/{encode_path_segment(address)}"
        return parse_object(self._transport.request("GET", path), Claimable.from_dict)

File: sdk/rust/src/api/distributor.rs
//! The `/distributor/*` endpoints: pot, shares and the public payout ledger.

use crate::client::{seg, FyuzClient, QueryPairs};
use crate::error::Result;
use crate::models::{
    AddressPayouts, Claimable, PayoutStats, Pot, RoundDetail, RoundReceipt, Shares,
};
use crate::params::{RoundsParams, SharesParams, ToQuery};

/// The Distributor endpoints, reached with
/// [`FyuzClient::distributor`](crate::FyuzClient::distributor).
///
/// The Distributor pays back a share of platform fees every round: most of the
/// pot pro-rata by points, the remainder to one VRF-picked winner. Everything
/// here is a public read of that ledger.
///
/// **Every `*_wei` field is an exact decimal string.** Lifetime totals run past
/// `2^53`, so parsing them into `f64` corrupts them silently. See the
/// [`models`](crate::models) docs.
///
/// ```no_run
/// # async fn demo() -> Result<(), fyuz::Error> {
/// let client = fyuz::FyuzClient::new()?;
/// let stats = client.distributor().get_stats().await?;
/// println!("{} wei paid out over {} rounds", stats.total_paid_wei, stats.rounds_settled);
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, Copy)]
pub struct Distributor<'a> {
    client: &'a FyuzClient,
}

impl<'a> Distributor<'a> {
    pub(crate) fn new(client: &'a FyuzClient) -> Self {
        Self { client }
    }

    /// `GET /distributor/stats` — lifetime payback totals.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let stats = client.distributor().get_stats().await?;
    /// // Decimal string, not a number: print it, do not parse it.
    /// println!("largest single payout: {} wei", stats.largest_payout_wei);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_stats(&self) -> Result<PayoutStats> {
        self.client
            .get_json("/distributor/stats", QueryPairs::new())
            .await
    }

    /// `GET /distributor/pot` — the live undistributed pot and the parameters
    /// needed to price a point.
    ///
    /// [`Pot::pot_bnb`] and [`Pot::total_points`] are `None` when the value could
    /// not be read. That is **unknown**, not zero — hide the figure rather than
    /// showing `0`.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let pot = client.distributor().get_pot().await?;
    /// if let (Some(bnb), Some(points)) = (pot.pot_bnb, pot.total_points) {
    ///     println!("{:.6} BNB per point", bnb * pot.distribute_bps / 10_000.0 / points);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_pot(&self) -> Result<Pot> {
        self.client
            .get_json("/distributor/pot", QueryPairs::new())
            .await
    }

    /// `GET /distributor/shares` — the share allocation for the current window,
    /// plus the exact calldata the round-runner posts on-chain.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::SharesParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let shares = client.distributor().get_shares(&SharesParams::new().limit(50)).await?;
    /// println!("{} holders, calldata {} bytes", shares.holders.len(), shares.packed.len() / 2 - 1);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_shares(&self, params: &SharesParams) -> Result<Shares> {
        self.client
            .get_json("/distributor/shares", params.to_query())
            .await
    }

    /// `GET /distributor/rounds` — the public payout ledger, newest first.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::RoundsParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// for round in client.distributor().list_rounds(&RoundsParams::new().limit(5)).await? {
    ///     // `None` means "not indexed yet", never zero.
    ///     match round.distributed_wei.as_deref() {
    ///         Some(wei) => println!("round {}: {wei} wei", round.round_id),
    ///         None => println!("round {}: pending settlement", round.round_id),
    ///     }
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn list_rounds(&self, params: &RoundsParams) -> Result<Vec<RoundReceipt>> {
        self.client
            .get_json("/distributor/rounds", params.to_query())
            .await
    }

    /// `GET /distributor/rounds/{id}` — one round's receipt and its recipient
    /// list.
    ///
    /// The receipt fields are flattened into the top level of the JSON object;
    /// in Rust they live under [`RoundDetail::round`].
    ///
    /// Answers `404` for an unknown round.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let detail = client.distributor().get_round(7).await?;
    /// println!("round {} paid {} wallets", detail.round.round_id, detail.payouts.len());
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_round(&self, id: i64) -> Result<RoundDetail> {
        self.client
            .get_json(&format!("/distributor/rounds/{id}"), QueryPairs::new())
            .await
    }

    /// `GET /distributor/payouts/{address}` — every payout one wallet has
    /// received.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let paid = client
    ///     .distributor()
    ///     .get_payouts("0x1111111111111111111111111111111111111111")
    ///     .await?;
    /// println!("{} wei across {} rounds", paid.total_wei, paid.rounds_paid);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_payouts(&self, address: &str) -> Result<AddressPayouts> {
        self.client
            .get_json(
                &format!("/distributor/payouts/{}", seg(address)),
                QueryPairs::new(),
            )
            .await
    }

    /// `GET /distributor/claimable/{address}` — wei owed from payout pushes the
    /// contract could not complete.
    ///
    /// [`Claimable::claimable_wei`] is `None` when the RPC is unreachable or no
    /// Distributor is configured. Hide the banner in that case rather than
    /// telling someone they are owed nothing.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let c = client
    ///     .distributor()
    ///     .get_claimable("0x1111111111111111111111111111111111111111")
    ///     .await?;
    ///
    /// match c.claimable_wei.as_deref() {
    ///     Some("0") => println!("nothing pending"),
    ///     Some(wei) => println!("{wei} wei claimable from {:?}", c.distributor_address),
    ///     None => println!("unknown — do not render a zero here"),
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_claimable(&self, address: &str) -> Result<Claimable> {
        self.client
            .get_json(
                &format!("/distributor/claimable/{}", seg(address)),
                QueryPairs::new(),
            )
            .await
    }
}
